### DISCUSSION TITLE
Is it possible to make release notes tool to consider the environment?

### DISCUSSION THREAD

**Discussion by bhirnyk** [rfennell/AzurePipelines#1566] [Q&A] [answered]
Hello @rfennell,
First of all thank you for the wonderfull Release Notes generation tool.
I have a CD pipeline that deploys app to different environments using [deployment job](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/deployment-jobs?view=azure-devops). 
Currently it looks like _XplatGenerateReleaseNotes@4_ is not considering the environment when it looks for the previous successful run. 
Is it possible to address this issue?

**Comment by rfennell** [comment]
If the correct parameters are set the task considers the stage to look for the last successful run

Have a look at [wiki](https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frfennell%2FAzurePipelines%2Fdiscussions%2F1566&data=05%7C01%7C%7C5e2eb680446c4c3ecc4b08dbea95441e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638361699029187867%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=NdkCHsUx0Achv5xAfCWaTBvvs4eAdlavdg%2FXeaPrbLQ%3D&reserved=0) for more details

**Reply by bhirnyk** [reply]
@rfennell  if you meant [this page](https://github.com/rfennell/AzurePipelines/wiki/GenerateReleaseNotes---Node-based-Cross-Platform-Task) - I've checked it. But the only parameter that seems applicable for my situation is _overrideBuildReleaseId_. But in this case I need to first somehow determine that previous build and it looks like non-trivial task.
Could you give another hint on this?

**Reply by rfennell** [reply, marked as answer]
You could use `overrideBuildReleaseId` but that in 'hard codes' the historic build to compare against. So does need some manual management as time goes on as you will wish to compare against new builds as the project proceeds

The section I meant was the 2nd option

| Option | Multi Stage YAML | Classic Build/Release |
|-|-|-|
| Generate notes for just the current build | Requires `checkstages=false` parameter| Run inside the build |
| Generate notes since the last successful release.  <br>Option 1. Place the task in a stage that is only run when you wish to generate release notes. Usually this will be guarded by branch based filters or manual approvals.  |  Requires `checkstages=true` parameter | Run inside the release. Supported and you can override the stage name used for comparison using the `overrideStageName` parameter
| Generate notes since the last successful release.  <br>Option 2. Set the task to look back for the last successful build that has a given tag |  Requires `checkstages=true` and the `tags` parameters| Not supported
| Generate notes since the last successful release.  <br>Option 3. Override the build that the task uses for comparison with a fixed value |  Requires `checkstages=true` and the `overrideBuildReleaseId` parameters | Run inside the release. Requires the `overrideBuildReleaseId` parameter

You set `checkstages=true` so it looks for historic runs where the current stage completed without error. You can supplement this with `overrideStageName` if you want to check for successful completion of the stage other than the current one the task is in.

The way we do it is have  stage dedicated to release notes generation that is only run, either manually or via branch protection rules in certain circumstances e.g. deployed to Live

I go through all of this in [this community conference session video](https://www.youtube.com/watch?v=xaV3dFoQdV8&t=332s) that might help

**Reply by bhirnyk** [reply]
Approach with tags worked fine for me, thanks!
